Issue #6 Create settings classes and add quiz settings to rule.php

andrewmadden opened the pull request.

andrewmadden commented:
Still pretty rough - Some feedback about structure would be helpful.

andrewmadden committed:
Issue #6 Add config_settings for handling settings per quiz, and settings manager for handling form elements

andrewmadden committed:
Issue #6 Create settings section for plugin in quiz settings and allow saving and deletion of settings per quiz

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public static function with_quizid(string $quizid) {
```
As per  https://docs.moodle.org/dev/Coding_style#Functions_and_Methods "Verbosity is encouraged: function names should be as illustrative as is practical to enhance understanding." Would suggest to pick a new name.

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+     */
+    public function exists() {
+        // Cast settings to array as empty stdClass will return true.
+        return !empty((array) $this->settings);
```
What would happen if I create a class with bunch of random settings? Do we need to hardcore settings we support (e.g. settings we  have fields for in DB)?

dmitriim replied:
Ah I see you have get_required_settings_keys! :+1:

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+                }
+            }
+        } catch (\dml_exception $e) {
+            debugging('quizaccess_seb - save settings: ' . $e->getMessage());
```
I think we should throw an exception here and not suppress it

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    public function delete_settings() : bool {
```
looking at the names of methods... as long as we are in the class already, it feels like we can drop "_settings" postfix/ E.g. delete_settings() -> delete()

dmitriim commented on rule.php:
```diff
@@ -62,6 +65,39 @@ public static function make (quiz $quizobj, $timenow, $canignoretimelimits) {
+        $existingsettings = $DB->get_record('quizaccess_seb_quizsettings', ['quizid' => $quizid], '*');
+
+        // Insert all the form elements before the 'security' section as a group.
+        foreach ($elements as $name => $type) {
```
it feels like we can just $settingsmanager->get_quiz_element_types() here and remove $elements

dmitriim commented on rule.php:
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
      *      which is the id of the quiz being saved.
      */
     public static function save_settings($quiz) {
+        $settingsmanager = new settings_manager();
```
can we abstract this please and have something like self::get_setting_manager()

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
+        } else {
+            $this->settings = $this->merge_new_and_existing_settings($this->strip_prefix($settings), $existingsettings);
```
I can't see that you validate incoming $settings. Maybe it's worth to spend some time and look at https://docs.moodle.org/dev/Persistent and rewrite the code? it seems like it has out of the box validation.

dmitriim replied:
or should __construct be private and we build the class only with "with_quizid" and "with_form_data" (where data is validated by the form) ?

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+     *
+     * @return string Item ID of stored file.
+     */
+    public function get_config_itemid() {
```
let's remove this code for now

dmitriim commented on rule.php:
```diff
@@ -62,6 +65,39 @@ public static function make (quiz $quizobj, $timenow, $canignoretimelimits) {
      */
     public static function add_settings_form_fields(mod_quiz_mod_form $quizform, MoodleQuickForm $mform) {
+        global $DB, $USER;
+        $settingsmanager = new settings_manager();
```
Love how you abstract all fields in the manager. :100:  Made usually messy form code more readable

dmitriim commented on classes/settings_manager.php:
```diff
@@ -0,0 +1,217 @@
+
+defined('MOODLE_INTERNAL') || die();
+
+class settings_manager {
```
Is there any reason why it's 'manager'? will it have actually manager functionality?

dmitriim commented on classes/settings_manager.php:
```diff
@@ -0,0 +1,217 @@
+     * @param bool $strippedprefix Defaults to true.
+     * @return array
+     */
+    public function get_file_picker_elements($strippedprefix = true) : array {
```
Lat's remove this code as it's not getting used.

dmitriim commented on classes/settings_manager.php:
```diff
@@ -0,0 +1,217 @@
+            if ($type == 'filepicker') {
+                if ($strippedprefix) {
+                    // Strip key prefix.
+                    $name = preg_replace("/^seb_/", "", $name);
```
I think we do that in other place, maybe abstract to a method?

dmitriim commented on rule.php:
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
+        $settingsmanager = new settings_manager();
+
+        $defaults = $settingsmanager->get_quiz_defaults();
+        // Set each required setting to default it not provided.
```
typo? " it not provided" -> " if not provided"

dmitriim commented on rule.php:
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
+        }
+
+        // Save/update settings from form.
+        config_settings::with_form_data($quiz)->save_settings();
```
Can't see where we validate incoming text data. E.g. "seb_expressionsblocked" or "seb_expressionsallowed". Should we validate them? If so then create a card for that. We will treat it as a feature and do later.

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+    public function __construct(int $quizid, \stdClass $settings = null) {
+        global $DB;
+
+        $this->quizid = $quizid;
```
Should we check if quiz with that ID exists at all?

dmitriim commented on classes/config_settings.php:
```diff
@@ -0,0 +1,259 @@
+
+        // If new settings are empty, set existing settings, otherwise merge existing settings with new settings.
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
```
As settings could be potentially empty. Do we want to update DB definition and have default values of required fields?

dmitriim replied:
Ah I can see that you validate it later. But still doesn't check if it matches DB type.

andrewmadden committed:
issue #6 renamed managed to provider and stripped it to just provide quiz setting info

andrewmadden committed:
Issue #6 : Swapped custom modal class to persistent class for quiz settings. Updated rule and tests to reflect change

andrewmadden commented:
* Now using a persistent class to represent the quiz settings. 
* Updated rule to use the persistent object.
* Renamed the `settings_manager` to `settings_provider` and stripped it down so it just provides the quiz settings info.
* Added tests for rule.php.
* Removed tests for `config_settings` and `settings_manager` because they are covered by persistent now.

andrewmadden committed:
Issue #6 Add tests to verify that the settings_provider data is sane

andrewmadden committed:
Issue #6 Remove double blank line to make code sniffer happy
